File: backend/models/schemas.py
from pydantic import BaseModel, Field, HttpUrl
from typing import Optional, List
from datetime import datetime
from enum import Enum

class SeverityLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

class ThreatType(str, Enum):
    PHISHING = "phishing"
    MALWARE = "malware"
    SPAM = "spam"
    CREDENTIAL_HARVESTING = "credential_harvesting"
    ZERO_DAY = "zero_day"

class UrlAnalysisPart(BaseModel):
    part: str
    value: str
    state: str
    note: Optional[str] = None

class AiReasoningItem(BaseModel):
    score: str
    state: str
    title: str
    desc: str

class IncidentReportBase(BaseModel):
    target_url: HttpUrl
    brand_impersonated: Optional[str] = "Unknown"
    source_ip: Optional[str] = None
    location: Optional[str] = None
    
class IncidentReportCreate(IncidentReportBase):
    pass

class IncidentReport(IncidentReportBase):
    id: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    type: ThreatType
    status: str = "Blocked"
    severity: SeverityLevel
    threat_score: float = Field(..., ge=0, le=100)
    url_analysis: List[UrlAnalysisPart] = []
    ai_reasoning: List[AiReasoningItem] = []

    class Config:
        orm_mode = True

class ModelPerformanceMetrics(BaseModel):
    version: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    accuracy: float
    loss: float
    training_data_size: int

class BlockedAttemptsStat(BaseModel):
    name: str # e.g., 'Mon', 'Tue' or actual date
    phishing: int = 0
    malware: int = 0
    spam: int = 0

File: server/src/models/schemas.js
const VALID_SEVERITY = new Set(["low", "medium", "high", "critical"]);
const VALID_THREAT_TYPES = new Set([
  "phishing",
  "malware",
  "spam",
  "credential_harvesting",
  "zero_day",
]);

export function validateIncident(payload) {
  const errors = [];

  if (!payload || typeof payload !== "object") {
    return ["Request body must be a JSON object."];
  }

  if (!payload.type || !VALID_THREAT_TYPES.has(payload.type)) {
    errors.push("type must be one of phishing, malware, spam, credential_harvesting, zero_day.");
  }

  if (!payload.severity || !VALID_SEVERITY.has(payload.severity)) {
    errors.push("severity must be one of low, medium, high, critical.");
  }

  if (typeof payload.threatScore !== "number" || payload.threatScore < 0 || payload.threatScore > 100) {
    errors.push("threatScore must be a number between 0 and 100.");
  }

  if (!payload.targetUrl || typeof payload.targetUrl !== "string") {
    errors.push("targetUrl is required and must be a string.");
  }

  return errors;
}